### PR TITLE
feat: add render template API support

### LINKED ISSUE #921: [Meilisearch v1.48.0] Add render template (experimental)
## Context

[Meilisearch 1.48.0](https://github.com/meilisearch/meilisearch/releases/tag/v1.48.0) introduces the render template route as an experimental feature.

## Objective

Update the SDK to support the [render template](https://www.meilisearch.com/docs/reference/api/template/render-documents-with-post) route as an experimental feature. Experimental features should require opt-in by the user.

## Tasks

- [ ] Add method to handle interacting with the render template API endpoint
- [ ] Add test cases for the new method
- [ ] Add a code sample in `.code-samples.meilisearch.yaml`, it should be the equivalent of the [documentation's CURL example](https://github.com/meilisearch/documentation/blob/ea674f4b86c52675c7996d817ec1b339c164dcff/.code-samples.meilisearch.yaml#L1327-L1340).

### PR DESCRIPTION
**Pull request by wayosu**
## Changes

Add render template API support for Meilisearch v1.48.0+ (Issue #921).

### What's included:
- **`TemplateRenderQuery`** — Builder-pattern request DTO with `setTemplate()` and `setInput()` methods
- **`TemplateRenderResult`** — Response DTO with `getTemplate()` and `getRendered()` accessors
- **`Templates` endpoint** — `POST /render-template` endpoint class
- **`HandlesTemplates` trait** — Delegate trait adding `renderTemplate()` convenience method to Client
- **Integration test** — Enables `renderRoute` experimental feature, tests inline template rendering
- **Code sample** — Added to `.code-samples.meilisearch.yaml`

### Pattern:
Follows the existing `DynamicSearchRules` pattern exactly — Contract DTOs → Endpoint class → Delegate trait → Client registration.

### Testing:
```php
$query = (new TemplateRenderQuery())
    ->setTemplate('inlineDocumentTemplate', '{{ doc.breed }} called {{ doc.name }}')
    ->setInput('inlineDocument', ['breed' => 'Jack Russell', 'name' => 'Iko']);

$response = $client->renderTemplate($query);
// $response->getTemplate() === '{{ doc.breed }} called {{ doc.name }}'
// $response->getRendered() === 'Jack Russell called Iko'
```

Closes #921

## Summary by CodeRabbit

* **New Features**
  * Added support for rendering templates through the client.
  * Introduced a new request/response flow for template rendering, including inline templates and optional input data.
  * Added a sample showing how to enable the experimental feature and render a template.

* **Tests**
  * Added coverage for template rendering behavior, including null input handling and request formatting.

### CODE REVIEW

**Comment by norkunas** [comment]
from release info:
> If input is null in the request, then rendered is null in the response

currently it's not possible, right?

**Comment by norkunas** [comment]
also:
```json
{
  "template": {
    "kind": "documentTemplate",
    "indexUid": "movies",
    "embedder": "myMoviesEmbedder"
  },
  "input": {
    "kind": "indexDocument",
    "indexUid": "movies",
    "id": "2"
}
```

how you would make this request?

File: tests/Endpoints/TemplatesTest.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+
+final class TemplatesTest extends TestCase
+{
+    protected function setUp(): void
+    {
+        parent::setUp();
+
+        $http = new Client($this->host, getenv('MEILISEARCH_API_KEY'));
+        $http->patch('/experimental-features', ['renderRoute' => true]);
+    }
```
[the remark this answers is not in the document]
Comment (Strift) [reply tests/Endpoints/TemplatesTest.php:19]: Existing tests might not do it, but I think it's good practice to do so 👍

**Review by Strift — CHANGES_REQUESTED** [review]
Hey, thanks for your PR! 

The CI is not passing. I suggest running linter + formatter + tests locally before requesting another review 🙏

PS. don't forget to check @norkunas and the AI's feedback!

**Comment by wayosu** [comment]
Hey @Strift, thanks for the review! All feedback has been addressed:

### ✅ @norkunas's feedback
- **Null input support**: `TemplateRenderQuery::setInput()` now accepts `?array`. When `setInput(null)` is called, the payload sends `"input": null` → API returns `rendered: null`. When not called at all, `input` is omitted from the request.
- **Flexible template/input kinds**: `setTemplate()` and `setInput()` now accept full array shapes, supporting both inline and document/index kinds:
  ```php
  // Document template + index document input
  $query->setTemplate(['kind' => 'documentTemplate', 'indexUid' => 'movies', 'embedder' => 'myEmbedder'])
        ->setInput(['kind' => 'indexDocument', 'indexUid' => 'movies', 'id' => '2']);
  ```

### ✅ CodeRabbit's AI feedback
- Code sample now includes `renderRoute` experimental-feature opt-in
- `setInput()` type narrowed from `mixed` to `?array`
- `tearDown()` added to reset `renderRoute` after tests

### ✅ CI
- `composer lint`: 0/181 files
- `composer phpstan`: no new errors (33 pre-existing in other files)
- Added 4 contract tests covering all template/input combinations + null input edge cases

Ready for re-review! 🙏

File: .hermes/plans/2026-06-30_031600-render-template-api.md
Comment (norkunas) [inline .hermes/plans/2026-06-30_031600-render-template-api.md:1]: this does not belong here

**Comment by wayosu** [comment]
Removed — sorry about that, was a local tooling artifact that shouldn't have been committed.

**Comment by wayosu** [comment]
@norkunas removed in 666ca08 + added .hermes/ to .gitignore. Thanks for catching that — won't happen again.

File: .gitignore [hunk 2]
```diff
@@ -5,3 +5,4 @@
 composer.lock
 phpunit.xml
 phpstan.neon
+.hermes/
```
Comment (norkunas) [inline .gitignore]: well it's not meilisearch-php generated dir, so it should go to your global .gitignore otherwise we'll need to ignore for all agents later

**Comment by wayosu** [comment]
@norkunas fair point — moved `.hermes/` to my global gitignore instead. Removed from project `.gitignore` in 60e0753.

**Comment by wayosu** [comment]
@norkunas — I wanted to apologize properly for the two issues in this PR:

1. **The `.hermes/` file** — that was a personal tooling artifact that never should have been committed. I've since configured my global gitignore properly so this won't happen again in any future contribution.

2. **The `.gitignore` change** — you're absolutely right that project repos shouldn't carry contributor-specific tooling ignores. That was shortsighted on my part.

Both are fixed in the latest commits. I appreciate you taking the time to review and correct this — it's a learning experience I won't repeat. Thanks for your patience.

File: src/Contracts/TemplateRenderQuery.php [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+declare(strict_types=1);
+
+namespace Meilisearch\Contracts;
+
+class TemplateRenderQuery
+{
+    /**
+     * @var array{kind: string, ...}
+     */
+    private array $template = ['kind' => '', 'inline' => ''];
```
Comment (norkunas) [inline src/Contracts/TemplateRenderQuery.php]: i'm not sure that it's useful to initialize this query with a wrong state, because `kind` is required right?
i think this query should accept both params via constructor

File: src/Contracts/TemplateRenderQuery.php [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+        return $this;
+    }
+
+    /**
+     * Set the input document for template rendering.
+     *
+     * Supports two kinds:
+     * - inlineDocument: ['kind' => 'inlineDocument', 'inline' => ['name' => 'John']]
+     * - indexDocument: ['kind' => 'indexDocument', 'indexUid' => 'movies', 'id' => '2']
```
Comment (norkunas) [inline src/Contracts/TemplateRenderQuery.php:57]: if there are only these two options then it could be properly documented via phpdoc:
```
@param array{kind: 'inlineDocument', inline: array<string, mixed>}|array{kind: 'indexDocument', indexUid: non-empty-string, id: string|int}|null $input
```

File: src/Contracts/TemplateRenderQuery.php [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+    private bool $inputSet = false;
+
+    /**
+     * Set the template to render.
+     *
+     * Supports two kinds:
+     * - inlineDocumentTemplate: ['kind' => 'inlineDocumentTemplate', 'inline' => '{{ doc.name }}']
+     * - documentTemplate: ['kind' => 'documentTemplate', 'indexUid' => 'movies', 'embedder' => 'myEmbedder']
+     *   or ['kind' => 'documentTemplate', 'indexUid' => 'movies', 'templateUid' => 'myTemplate']
```
Comment (norkunas) [inline src/Contracts/TemplateRenderQuery.php:41]: if there are only these three options then it could be properly documented via phpdoc:
```
@param array{kind: 'inlineDocumentTemplate', inline: string}|array{kind: 'documentTemplate', indexUid: non-empty-string, embedder: non-empty-string}|array{kind: 'documentTemplate', indexUid: non-empty-string, templateUid: non-empty-string} $template
```

**Comment by wayosu** [comment]
@norkunas thanks for the review! I've addressed all three points:

1. **Constructor-based initialization** — `TemplateRenderQuery` now takes `$template` (required) and `$input` (optional) via constructor, eliminating the invalid default state. The builder setters are still available for modification after construction.

2. **Precise PHPStan union types** — `setTemplate()`, `setInput()`, and `toArray()` now use the exact union array shapes you suggested for all template and input variants.

3. Also updated tests and the code sample to use the new constructor syntax.

All lint, PHPStan, and tests pass locally. Ready for re-review when you have a chance.
